src/super_gradients/training/kd_model/kd_model.py

```diff
@@ -18,101 +19,194 @@ class KDModel(SgModel):
 
-          kd_model_external_checkpoint_path: The path to the external checkpoint to be loaded. Can be absolute or relative
-                                           (ie: path/to/checkpoint.pth). If provided, will automatically attempt to
-                                           load the checkpoint even if the load_checkpoint flag is not provided.
-                                           (deafult=None)
         """
+        kwargs.setdefault("student_architecture", None)
+        kwargs.setdefault("teacher_architecture", None)
+        kwargs.setdefault("student_arch_params", {})
+        kwargs.setdefault("teacher_arch_params", {})
+        kwargs.setdefault("run_teacher_on_eval", False)
 
-        student_net = get_param(arch_params, "student")
-        teacher_net = get_param(arch_params, "teacher")
+        self._validate_args(arch_params, architecture, checkpoint_params, **kwargs)
 
-        student_architecture = get_param(arch_params, "student_architecture")
-        teacher_architecture = get_param(arch_params, "teacher_architecture")
-        student_arch_params = get_param(arch_params, "student_arch_params")
-        teacher_arch_params = get_param(arch_params, "teacher_arch_params")
-        teacher_checkpoint_path = get_param(arch_params, "teacher_checkpoint_path")
+        super(KDModel, self).build_model(architecture=architecture, arch_params=arch_params,
+                                         checkpoint_params=checkpoint_params, **kwargs)
 
-        if (student_net and student_architecture) or (not student_net and not student_architecture):
-            raise ValueError("Exactly one of: student, student_architecture should be passed through arch_params")
-        if (teacher_net and teacher_architecture) or (not teacher_net and not teacher_architecture):
-            raise ValueError("Exactly one of: teacher, teacher_architecture should be passed through arch_params")
+    def _validate_args(self, arch_params, architecture, checkpoint_params, **kwargs):
+        student_architecture = get_param(kwargs, "student_architecture")
+        teacher_architecture = get_param(kwargs, "teacher_architecture")
+        student_arch_params = get_param(kwargs, "student_arch_params")
+        teacher_arch_params = get_param(kwargs, "teacher_arch_params")
 
-        # DERIVE NUMBER OF CLASSES FROM DATASET INTERFACE IF NOT SPECIFIED
-        if 'num_classes' not in student_arch_params.keys():
-            if self.dataset_interface is None:
-                raise Exception('Error', 'Number of classes not defined in students arch params and dataset is not '
-                                         'defined')
-            else:
-                student_arch_params['num_classes'] = len(self.classes)
+        if get_param(checkpoint_params, 'pretrained_weights') is not None:
+            raise ValueError("pretrained_weights is ambiguous for KD models.")
```

Comment:
@shaydeci  - Great work on all of the exception handling!

I want to point out that all over this code you are using vanilla exceptions... We should consider using custom exceptions since their uniqueness allows the stack trace to point to the location in the codebase easily.

For example:
Creating a base class for exception (KDModelException that inherits from Exception) will allow the user to search for it in the code base and find its source file... 
Then we can inherit from this base exception class and raise more specific exceptions for eachof the relevant use cases...